src/ray/gcs/gcs_server/gcs_resource_report_poller.cc

```diff
@@ -0,0 +1,149 @@
+
+  nodes_[node_id] = state;
+  to_pull_queue_.push_front(state);
+  RAY_LOG(DEBUG) << "Node was added with id: " << node_id;
+
+  polling_service_.post([this]() { TryPullResourceReport(); });
+}
+
+void GcsResourceReportPoller::HandleNodeRemoved(
+    const std::shared_ptr<rpc::GcsNodeInfo> &node_info) {
+  NodeID node_id = NodeID::FromBinary(node_info->node_id());
+  {
+    absl::MutexLock guard(&mutex_);
+    nodes_.erase(node_id);
+    RAY_CHECK(!nodes_.count(node_id));
+  }
+  RAY_LOG(DEBUG) << "Node removed (node_id: " << node_id
+                 << ")# of remaining nodes: " << nodes_.size();
+}
+
+void GcsResourceReportPoller::Tick() { TryPullResourceReport(); }
+
+void GcsResourceReportPoller::TryPullResourceReport() {
+  absl::MutexLock guard(&mutex_);
+  int64_t cur_time = get_current_time_milli_();
+
+  RAY_LOG(DEBUG) << "Trying to pull inflight_pulls " << inflight_pulls_ << "/"
+                 << max_concurrent_pulls_ << ", queue size: " << to_pull_queue_.size();
+  while (inflight_pulls_ < max_concurrent_pulls_ && !to_pull_queue_.empty()) {
+    auto to_pull = to_pull_queue_.front();
+    if (cur_time < to_pull->next_pull_time) {
+      break;
+    }
+
+    to_pull_queue_.pop_front();
+
+    if (!nodes_.count(to_pull->node_id)) {
+      RAY_LOG(DEBUG)
+          << "Update finished, but node was already removed from the cluster. Ignoring.";
+      continue;
+    }
+
+    PullResourceReport(to_pull);
+  }
+}
+
+void GcsResourceReportPoller::PullResourceReport(const std::shared_ptr<PullState> state) {
+  inflight_pulls_++;
+
+  request_report_(
+      state->address, raylet_client_pool_,
+      [this, state](const Status &status, const rpc::RequestResourceReportReply &reply) {
+        if (status.ok()) {
+          // TODO (Alex): This callback is always posted onto the main thread. Since most
+          // of the work is in the callback we should move this callback's execution to
+          // the polling thread. We will need to implement locking once we switch threads.
+          handle_resource_report_(reply.resources());
+        } else {
+          RAY_LOG(INFO) << "Couldn't get resource request from raylet " << state->node_id
+                        << ": " << status.ToString();
+        }
+        polling_service_.post([this, state]() { NodeResourceReportReceived(state); });
+      });
+}
+
+void GcsResourceReportPoller::NodeResourceReportReceived(
+    const std::shared_ptr<PullState> state) {
+  absl::MutexLock guard(&mutex_);
+  inflight_pulls_--;
+  NodeID &node_id = state->node_id;
+  if (nodes_.count(node_id) == 0) {
+    RAY_LOG(INFO)
+        << "Resource report received, but the node is no longer in the cluster. NodeID: "
+        << node_id;
+  } else {
+    state->next_pull_time = get_current_time_milli_() + poll_period_ms_;
+    to_pull_queue_.push_back(state);
+  }
+
+  polling_service_.post([this] { TryPullResourceReport(); });
```

Comment:
Why do we need this post, isn't there a Tick() scheduled already?

Reply:
I think we explicitly want to pull more things without waiting for a tick. cc @stephanie-wang

Reply:
Why not just wait for the tick? Especially if it's firing at high frequency (<10ms) We should either have a ticker or call it periodically, having both is unnecessary.

Reply:
I don't think it's going to make a big difference either way, but why not do it this way since it's more responsive? 

The only scenario I can see this mattering in is if rpc's fail very quickly for some reason.